cpp/.clang-format

```diff
@@ -0,0 +1,156 @@
+---
+# Refer to the following link for the explanation of each params:
+#   http://releases.llvm.org/8.0.0/tools/clang/docs/ClangFormatStyleOptions.html
+Language: Cpp
+# BasedOnStyle: Google
+AccessModifierOffset: -1
+AlignAfterOpenBracket: Align
+AlignConsecutiveAssignments: true
+AlignConsecutiveDeclarations: false
+AlignEscapedNewlines: Left
+AlignOperands: true
+AlignTrailingComments: true
+AllowAllParametersOfDeclarationOnNextLine: true
+AllowShortBlocksOnASingleLine: true 
+AllowShortCaseLabelsOnASingleLine: true
+AllowShortFunctionsOnASingleLine: All
+AllowShortIfStatementsOnASingleLine: true
+AllowShortLoopsOnASingleLine: true
+# This is deprecated
+AlwaysBreakAfterDefinitionReturnType: All
+AlwaysBreakAfterReturnType: All
+AlwaysBreakBeforeMultilineStrings: true
+AlwaysBreakTemplateDeclarations: Yes
+BinPackArguments:  true       
+BinPackParameters: true
+BraceWrapping:
+  AfterClass:            false
+  AfterControlStatement: false
+  AfterEnum:             false
+  AfterFunction:         false
+  AfterNamespace:        false
+  AfterObjCDeclaration:  false
+  AfterStruct:           false
+  AfterUnion:            false
+  AfterExternBlock:      false
+  BeforeCatch:           false
+  BeforeElse:            false
+  IndentBraces:          false
+  # disabling the below splits, else, they'll just add to the vertical length of source files!
+  SplitEmptyFunction: false
```

Comment:
This one is an unrecognized key for me. As part of this PR,  can you work out how to ensure all of us have the same version of clang-format installed?

Reply:
You need to have Clang-8.0 (which is available due to cuML relying on it), I had the same problem at first. I will look into making sure that this is configured properly for everyone.

Reply:
```
sudo update-alternatives --install /usr/bin/clang clang /usr/bin/clang-8   81 --slave /usr/bin/clang++ clang++ /usr/bin/clang++-8    --slave /usr/share/man/man1/clang.1.gz clang.1.gz /usr/share/man/man1/clang-8.1.gz --slave /usr/bin/clang-tidy clang-tidy /usr/bin/clang-tidy-8  --slave /usr/bin/clang-format clang-format /usr/bin/clang-format-8
sudo update-alternatives --install /usr/bin/clang clang /usr/bin/clang-6.0 60 --slave /usr/bin/clang++ clang++ /usr/bin/clang++-6.0  --slave /usr/share/man/man1/clang.1.gz clang.1.gz /usr/share/man/man1/clang-6.0.1.gz --slave /usr/bin/clang-tidy clang-tidy /usr/bin/clang-tidy-6.0 --slave /usr/bin/clang-format clang-format /usr/bin/clang-format-6.0
```
Ubuntu has 6.0 as default. Install clang-8 and clang-format-8 and create update-alternatives entries using commands above.

Reply:
> can you work out how to ensure all of us have the same version of clang-format installed?

This will almost certainly have to be via a conda package.

Reply:
I don't know if conda has the entire `clang-tools` suite, but here's how I install `clang-tools-10` in ubuntu (works on 16 and 18):
```
release=$(lsb_release -cs)
wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
echo "deb http://apt.llvm.org/$release/ llvm-toolchain-$release main
deb-src http://apt.llvm.org/$release/ llvm-toolchain-$release main
" | sudo tee /etc/apt/sources.list.d/llvm.list

sudo apt update && sudo apt install -y clang-tools-10 clangd-10
sudo update-alternatives --install /usr/bin/clangd clangd /usr/bin/clangd-10 100
```

Clang 10 has much more support for CUDA 10+, and this'll give you the `clangd` language server used by the `vscode-clangd` intellisense plugin.

Reply:
Clang 10 isn't officially released yet AFAIK: http://releases.llvm.org/

There's Clang 9.0.0 conda packages: https://anaconda.org/conda-forge/clang-tools/files

Reply:
No it's not, but it's stable enough that it's been promoted to be the default packages in LLVM's apt repo (https://apt.llvm.org/).